### PR TITLE
spec: Propose AC Identifier Type. And redefine AC Name Type.

### PR DESCRIPTION
**Pull request by yifan-gu**
AC Identifier Type's format is the same as the old AC Name Type,
which contains "/" and ".", it is used for fields that have global
namespace, such as image name and image labels.

AC Name Type's removes "/" and "." from the AC Identifier Type.
This type is used for other fields that can be identified relatively
within a pod, image, or app context, such as app name, port name,
volume name.

### CODE REVIEW

**Comment by yifan-gu** [comment]
First, I am bad at naming.
I have thought about an alternative that create `AC Label Type` as for names that do not have `/` or `.`.
But then we will not be able to use `AC Label Type` for labels, so I gave up. Instead, I think maybe renaming the `AC Name` to something more `global` is more reasonable ?

**Comment by yifan-gu** [comment]
This tries to fix #342 #247 and partially #351
/cc @thockin @jonboulle @philips

**Comment by yifan-gu** [comment]
This tries to fix #342 #247 and partially #351
/cc @thockin @jonboulle @philips

**Comment by jonboulle** [comment]
How does this address #351?

**Comment by jonboulle** [comment]
Not sure on the naming front, will have to mull on it...

**Comment by yifan-gu** [comment]
@jonboulle I think as we now have two types, the decision of which type to use for labels implicitly answers #351  No?

**Comment by yifan-gu** [comment]
OK, its' not answering #351 in whole, it just marks we go with the direction of namespacing the labels or not.

**Comment by jonboulle** [comment]
@yifan-gu what do you think of altering this to cover #406?

**Comment by yifan-gu** [comment]
@jonboulle Yes, I think we just need to update the `AC Identifier` here

**Comment by thockin** [comment]
This looks like a good step to me.

**Comment by yifan-gu** [comment]
Updated:
AC Identifier: `^[a-z0-9]+([-._~/][a-z0-9]+)*$` (lowercase alphanumeric plus `- . _ ~ /`, must start with alphanumeric)
AC Name: `^[a-z0-9]+([-][a-z0-9]+)*$` (lowercase alphanumeric plus `- `, must start with aplhanumeric)

AC Identifier will be used for image names and label keys(see #351).
AC Name will be used in all other places.

**Comment by yifan-gu** [comment]
I will change the implementation code if this looks good to us for now :)

File: SPEC.md [hunk 1]
```diff
@@ -763,8 +779,8 @@ JSON Schema for the Image Manifest (app image manifest, ACI manifest), conformin
...
-* **labels** (list of objects, optional) used during image discovery and dependency resolution. The listed objects must have two key-value pairs: *name* is restricted to the [AC Name](#ac-name-type) formatting and *value* is an arbitrary string. Label names must be unique within the list, and (to avoid confusion with the image's name) cannot be "name". Several well-known labels are defined:
+* **name** (string, required) a human-readable name for this App Container Image (string, restricted to the [AC Identifier](#ac-identifier-type) formatting). This is not expected to be unique (see the **version** label) but SHOULD have a URL-like structure to facilitate **[App Container Image Discovery](#app-container-image-discovery)**. If this image is resolved through the discovery process, this field MUST match the name used for discovery.
```
Comment (jonboulle) [inline SPEC.md]: This is supposed to stay AC Name surely?

Comment (jonboulle) [inline SPEC.md]: er, ignore me, reading backwards

**Comment by jzelinskie** [comment]
sgtm, idk how this effects the discovery protocols though (i.e. escaping/encoding/decoding in order to include symbols like `-._~/`).

**Comment by jonboulle** [comment]
@jzelinskie why would we need to escape/decode any of those symbols?

**Comment by jonboulle** [comment]
@yifan-gu Sorry for the delay - this LGTM so I'd say go ahead with the schema changes.

**Comment by jonboulle** [comment]
cc @krobertson

File: SPEC.md [hunk 2]
```diff
@@ -552,19 +552,35 @@ Accessible at `$AC_METADATA_URL/acMetadata/v1/pod/hmac`
...
 
-### AC Name Type
+### AC Identifier Type
 
-An AC Name Type is restricted to lowercase characters accepted by the DNS [RFC1123](http://tools.ietf.org/html/rfc1123#page-13) and "/".
-An AC Name Type cannot be an empty string and must begin and end with an alphanumeric character.
-An AC Name Type will match the following [RE2](https://code.google.com/p/re2/wiki/Syntax) regular expression: `^[a-z0-9]+([-./][a-z0-9]+)*$`
+An AC Identifier Type is restricted to lowercase URI unreserved characters defined in [RFC3986](http://tools.ietf.org/html/rfc3986#section-2.3)
+An AC Identifier Type cannot be an empty string and must begin and end with an alphanumeric character.
+An AC Identifier Type will match the following [RE2](https://code.google.com/p/re2/wiki/Syntax) regular expression: `^[a-z0-9]+([-._~/][a-z0-9]+)*$`
```
Comment (thockin) [inline SPEC.md]: so `v~-_._-~v` is a valid identifier.  Interesting.  We specifically said that identifiers are a relatively loose string (like you spec) preceeded by an optional DNS name.  Specifically, the existence of a / at all implies that the first part is a DNS subdomain.  It's a little tighter than this - maybe this is OK.  Just go through and think about the most ridiculous valid string a user could input.

**Comment by jzelinskie** [comment]
@jonboulle I think it was simply confusion on my behalf. The current description for AC Identifier references the DNS RFC for its schema and I was under the impression that DNS was desired to be the "lowest common denominator" in terms of character sets such that a discovery protocol could be implemented on top of it.

File: SPEC.md (hunk 2 above)
Comment (yifan-gu) [inline SPEC.md]: @thockin Yeah, we need to address this... Thanks for catching!

**Comment by yifan-gu** [comment]
Another problem I hit during the change is for the name of isolators.
e.g. `resource/block-bandwidth`, `os/linux/capabilities-retain-set`, which container `/` in the name, so it will not satisfy the `ACName`, and we have to use `ACIdentifier`. How do you guys think about this?

File: SPEC.md (hunk 2 above)
Comment (jonboulle) [inline SPEC.md]: @thockin do you have any ideas for a sensible prescription here?

**Comment by jonboulle** [comment]
@yifan-gu sorry, you will need to rebase

`ACIdentifier` is fine for isolators since those are already hierarchically namespaced.

**Comment by yifan-gu** [comment]
@jonboulle I have rebased. Take a look?

**Comment by yifan-gu** [comment]
Ehh, I see, the spec is breaked into different sections..

**Comment by yifan-gu** [comment]
Now rebased.

**Comment by jonboulle** [comment]
LGTM, another pair of eyes would be good though.
@vbatts @krobertson

**Comment by jonboulle** [comment]
or @mpasternacki ?

File: spec/types.md [hunk 3]
```diff
@@ -2,19 +2,35 @@
...
+An AC Identifier Type is restricted to lowercase URI unreserved characters defined in [RFC3986](http://tools.ietf.org/html/rfc3986#section-2.3)
+An AC Identifier Type cannot be an empty string and must begin and end with an alphanumeric character.
+An AC Identifier Type will match the following [RE2](https://code.google.com/p/re2/wiki/Syntax) regular expression: `^[a-z0-9]+([-._~/][a-z0-9]+)*$`
 
 Examples:
 
 * database
 * example.com/database
 * example.com/ourapp
+* example.com/~user/app_v1
```
Comment (vbatts) [inline spec/types.md]: nice.

**Comment by vbatts** [comment]
So, this deprecates ACName, but it's still to be left as a type?

**Comment by vbatts** [comment]
no. The annotations will still use ac-name-type, and pod names.

I'm not opposed to this identifier, there is just enough overlap that having both seems confusing.

**Comment by yifan-gu** [comment]
@vbatts The original reason for having two types is that we want more strict names for those app names, port names etc. (in particular, we want to eliminate `/` `.` in these names #342)

**Comment by mpasternacki** [comment]
LGTM

**Comment by mpasternacki** [comment]
Though I'd nitpick about the name: I'd use new name for new type, especially that I'd expect "identifier" to be shorter than "name" – like "id" vs "full name". `ACIdentifier` is similar to current `ACName`, and proposed `ACName` is essentially a "short id" kind of type. I'd feel better about it if the names were reversed, but this is a cosmetic issue, so I won't argue about it.

**Comment by vbatts** [comment]
LGTM

**Pull request merged**